File: graph/Graph.java
package graph;
import java.util.ArrayList;
import java.util.List;
public abstract class Graph<T> implements IGraph<T> {
    protected final List<INode<T>> visited;
    public Graph(){
        this.visited = new ArrayList<>();
    }
    public Graph(List<INode<T>> newList){
        this.visited = newList;
    }
    @Override
    public void explore(INode<T> start){
        this.Reinitialize();
        this.Exploring(start);
    }
    protected abstract void Exploring(INode<T> node);
    protected void Reinitialize(){
        this.visited.clear();
    }
}

File: src/datastructures/graphs/Graph.php
<?php
declare(strict_types=1);
namespace ZimCodes\Self_Learning\DataStructures\Graphs;
abstract class Graph implements IGraph {
    protected array $visited;
    public function __construct(){
        $this->visited = [];
    }
    public function explore(INode $node): void{
        $this->reInit();
        $this->exploring($node);
    }
    protected function reInit(){
        $this->visited = [];
    }
    public function getPathString(): array
    {
        $arr = [];
        foreach ($this->visited as $node) {
            $arr[] = $node->getName();
        }
        return $arr;
    }

    abstract protected function exploring(INode $node):void;
    abstract public function getPath(): array;
}